node/narwhal/src/helpers/codec.rs

```diff
@@ -84,3 +84,311 @@ impl<N: Network> Decoder for EventCodec<N> {
+        }
+    }
+}
+
+impl NoiseState {
+    pub fn into_post_handshake_state(self) -> Self {
+        if let Self::Handshake(noise_state) = self {
+            let noise_state = noise_state.into_stateless_transport_mode().expect("handshake isn't finished");
+            Self::PostHandshake(PostHandshakeState { state: Arc::new(noise_state), tx_nonce: 0, rx_nonce: 0 })
+        } else {
+            panic!()
+        }
+    }
+}
+
+pub struct NoiseCodec<N: Network> {
+    codec: LengthDelimitedCodec,
+    event_codec: EventCodec<N>,
+    pub noise_state: NoiseState,
+}
+
+impl<N: Network> NoiseCodec<N> {
+    pub fn new(noise_state: NoiseState) -> Self {
+        Self { codec: LengthDelimitedCodec::new(), event_codec: EventCodec::default(), noise_state }
+    }
+}
+
+impl<N: Network> Encoder<EventOrBytes<N>> for NoiseCodec<N> {
+    type Error = std::io::Error;
+
+    fn encode(&mut self, message_or_bytes: EventOrBytes<N>, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let message_type = message_or_bytes.message_type();
+
+        let ciphertext = match self.noise_state {
+            NoiseState::Handshake(ref mut noise) => {
+                match message_or_bytes {
+                    // Don't allow message sending before the noise handshake has completed.
+                    EventOrBytes::Event(_) => unimplemented!(),
+                    EventOrBytes::Bytes(bytes) => {
+                        let mut buffer = [0u8; MAX_MESSAGE_LEN + 1];
+                        let len = noise
+                            .write_message(&bytes, &mut buffer[1..])
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        // Set the message type flag.
+                        buffer[0] = message_type as u8;
+
+                        buffer[..len + 1].into()
+                    }
+                }
+            }
+
+            NoiseState::PostHandshake(ref mut noise) => {
+                // Encode the message using the event codec.
+                let mut bytes = BytesMut::new();
+                match message_or_bytes {
+                    // Don't allow sending raw bytes after the noise handshake has completed.
+                    EventOrBytes::Bytes(_) => unimplemented!(),
+                    EventOrBytes::Event(event) => self.event_codec.encode(event, &mut bytes)?,
+                }
+
+                // Chunk the payload if necessary.
+                //
+                // A Noise transport message is simply an AEAD ciphertext that is less than or
+                // equal to 65535 bytes in length, and that consists of an encrypted payload plus
+                // 16 bytes of authentication data.
+                //
+                // See: https://noiseprotocol.org/noise.html#the-handshakestate-object
+                let chunked_plaintext_msg: Vec<_> = bytes.chunks(MAX_MESSAGE_LEN - 16).collect();
```

Comment:
same note as on decryption - check if `par_chunks` could be used here to avoid the intermediate vector